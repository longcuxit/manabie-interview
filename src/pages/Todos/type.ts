import { TodoModel, TodoStatus } from "models/Todo";
import { FocusEventHandler } from "react";

export type StatusFilter = TodoStatus | "ALL";

export interface TodoFilterState {
  keyword: string;
  status: StatusFilter;
}

export interface TodoFilterActions {
  keyword(keyword: string): void;
  status(status: StatusFilter): void;
}

export interface TodoItemProps {
  item: TodoModel;
}

export interface TodoFormProps {
  defaultValue?: string;
  onSubmit: (value: string) => Promise<void>;
  onBlur?: FocusEventHandler<HTMLFormElement>;
}
