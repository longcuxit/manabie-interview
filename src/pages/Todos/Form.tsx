import classNames from "classnames";
import { useLoading } from "components/Loader";
import { FocusEventHandler, FormEvent, MouseEvent, useState } from "react";
import { InputGroup, FormControl, Spinner } from "react-bootstrap";
import { Send, PencilFill } from "react-bootstrap-icons";

export interface TodoFormProps {
  defaultValue?: string;
  onSubmit: (value: string) => Promise<void>;
  onBlur?: FocusEventHandler<HTMLFormElement>;
}

export default function TodoForm({
  defaultValue = "",
  onSubmit,
  onBlur,
}: TodoFormProps) {
  const [value, setValue] = useState(defaultValue);
  const [loading, pushLoader] = useLoading();

  const handleSubmit = async (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    if (!value) return;
    await pushLoader(onSubmit(value));
    setValue("");
  };
  const handleBtnFocus = ({ currentTarget }: MouseEvent<HTMLButtonElement>) => {
    const input = currentTarget.previousElementSibling as HTMLInputElement;
    input.focus();
  };

  return (
    <InputGroup
      as="form"
      className={classNames(loading && "disabled")}
      onSubmit={handleSubmit}
      onBlur={onBlur}
    >
      <FormControl
        autoFocus
        value={value}
        placeholder="What you will do ..."
        onChange={({ target }) => setValue(target.value.trimLeft())}
      />
      {loading ? (
        <InputGroup.Text>
          <Spinner animation="border" size="sm" />
        </InputGroup.Text>
      ) : value ? (
        <InputGroup.Text as="button" type="submit" className="btn-send">
          <Send />
        </InputGroup.Text>
      ) : (
        <InputGroup.Text
          as="button"
          className="btn-focus"
          type="button"
          onClick={handleBtnFocus}
        >
          <PencilFill className="icon-pencil-fill" />
        </InputGroup.Text>
      )}
    </InputGroup>
  );
}
