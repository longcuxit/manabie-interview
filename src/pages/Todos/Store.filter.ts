import { TodoStatus } from "models/Todo";
import { debounce } from "utils";
import { Store } from "utils/Store";

export type StatusFilter = TodoStatus | "ALL";

interface State {
  keyword: string;
  status: StatusFilter;
}

const initialState: State = {
  keyword: "",
  status: "ALL",
};

const store = new Store(initialState, ({ partial }) => ({
  keyword: debounce((keyword: string) => partial({ keyword }), 300),
  status: (status: StatusFilter) => partial({ status }),
}));

export const useTodoFilter = store.createHook();
