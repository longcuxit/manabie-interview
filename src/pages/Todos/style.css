.Page-todos .label-big-holder {
  margin: -8px;
  padding: 8px;
  cursor: pointer;
}
