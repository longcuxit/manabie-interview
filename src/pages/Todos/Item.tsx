import { FocusEvent, useState } from "react";
import { CloseButton, FormCheck, ListGroup, Spinner } from "react-bootstrap";

import classNames from "classnames";

import { TodoStatus } from "models/Todo";
import { useTodoActions } from "store/Todos";
import { useLoading, withLoader } from "components/Loader";
import { useAsyncConfirm } from "components/AsyncModal";
import { createLocalStorage } from "utils";
import TodoForm from "./Form";
import { TodoItemProps } from "./type";

const askMeDelete = createLocalStorage("DeleteAnItem", true);

function TodoItem({ item }: TodoItemProps) {
  const actions = useTodoActions();
  const showConfirm = useAsyncConfirm();
  const [loading, pushLoader] = useLoading();

  const [editMode, setEditMode] = useState(false);

  const isActive = item.status === TodoStatus.ACTIVE;

  const handleDelete = () => {
    return pushLoader(
      (async () => {
        const isRemove =
          !askMeDelete.get() ||
          (await showConfirm({
            title: "Delete todo",
            message: "Are you sure delete this Todo?",
            footer: (
              <FormCheck
                className="flex-fill"
                label="Don't ask again!"
                onChange={({ target }) => askMeDelete.set(!target.checked)}
              />
            ),
          }));
        if (isRemove) await actions.remove(item.id);
      })()
    );
  };

  const handleBlur = (e: FocusEvent<HTMLElement>) => {
    const related = e.relatedTarget as HTMLElement;
    if (related && related.parentElement === e.currentTarget) return;
    setEditMode(false);
  };

  return (
    <ListGroup.Item
      className={classNames(
        "d-flex flex-row p-2 align-items-center",
        loading && "disabled"
      )}
      onDoubleClick={() => setEditMode(true)}
    >
      {editMode ? (
        <TodoForm
          defaultValue={item.content}
          onBlur={handleBlur}
          onSubmit={async (value: string) => {
            await actions.updateContent(item.id, value);
            setEditMode(false);
          }}
        />
      ) : (
        <>
          <label className="label-big-holder">
            <FormCheck
              checked={!isActive}
              onChange={() =>
                pushLoader(actions.updateStatus(item.id, !isActive))
              }
            />
          </label>
          <div className="flex-fill px-2">{item.content}</div>
          {loading ? (
            <Spinner animation="border" size="sm" className="mx-1" />
          ) : (
            <CloseButton onClick={handleDelete} />
          )}
        </>
      )}
    </ListGroup.Item>
  );
}
export default withLoader(TodoItem);
