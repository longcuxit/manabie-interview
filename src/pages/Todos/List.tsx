import { useMemo } from "react";
import { Container, ListGroup } from "react-bootstrap";

import { useTodos } from "store/Todos";
import TodoItem from "./Item";
import { useTodoFilter } from "./Store.filter";

export default function TodoList() {
  const [{ todos }] = useTodos();
  const [{ status, keyword }] = useTodoFilter();

  const visible = useMemo(() => {
    let visible = Array.from(todos.values());

    if (status !== "ALL") {
      visible = visible.filter((todo) => todo.status === status);
    }

    if (keyword.length) {
      const lowerKey = keyword.toLowerCase();
      visible = visible.filter((todo) =>
        todo.content.toLowerCase().includes(lowerKey)
      );
    }

    return visible;
  }, [status, keyword, todos]);

  return (
    <Container className="my-2">
      <ListGroup className="ToDo__list">
        {visible.map((todo) => (
          <TodoItem key={todo.id} item={todo} />
        ))}
      </ListGroup>
    </Container>
  );
}
