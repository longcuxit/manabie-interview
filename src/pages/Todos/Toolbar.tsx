import { MouseEvent } from "react";
import {
  Navbar,
  Container,
  DropdownButton,
  Dropdown,
  Button,
  InputGroup,
  FormControl,
  FormCheck,
} from "react-bootstrap";
import { BoxArrowRight, Search, Trash, XLg } from "react-bootstrap-icons";

import { createLocalStorage } from "utils";
import { TodoStatus } from "models/Todo";
import { useAuthAction } from "store/Auth";
import { useTodos } from "store/Todos";

import { usePushLoader } from "components/Loader";
import { PartialCheckbox } from "components/PartialCheckbox";
import { useAsyncConfirm } from "components/AsyncModal";

import { useTodoFilter } from "./Store.filter";
import { StatusFilter } from "./type";

const statusList: StatusFilter[] = [
  "ALL",
  TodoStatus.ACTIVE,
  TodoStatus.COMPLETED,
];

const askMeDeleteAll = createLocalStorage("DeleteAllItems", true);

const LogoutButton = () => {
  const authAction = useAuthAction();
  return (
    <Button className="ms-2" size="sm" onClick={authAction.logout}>
      <BoxArrowRight />
    </Button>
  );
};

export default function TodoToolbar() {
  const showConfirm = useAsyncConfirm();
  const pushLoader = usePushLoader();

  const [{ todos, countActive }, todoActions] = useTodos();
  const [{ status, keyword }, filterActions] = useTodoFilter();

  const handleDelete = async () => {
    const isRemove =
      !askMeDeleteAll.get() ||
      (await showConfirm({
        title: "Delete all completed todo",
        message: "Are you sure to clear all completed Todo?",
        footer: (
          <FormCheck
            className="flex-fill"
            label="Don't ask again!"
            onChange={({ target }) => askMeDeleteAll.set(!target.checked)}
          />
        ),
      }));
    if (isRemove) pushLoader(todoActions.clearAll());
  };

  const handleClear = ({ currentTarget }: MouseEvent<HTMLButtonElement>) => {
    const input = currentTarget.previousElementSibling as HTMLInputElement;
    filterActions.keyword("");
    input.value = "";
  };

  const handleFocus = ({ currentTarget }: MouseEvent<HTMLButtonElement>) => {
    const input = currentTarget.previousElementSibling as HTMLInputElement;
    input.focus();
  };

  return (
    <Navbar bg="primary" className="sticky-top">
      <Container>
        {Boolean(todos.size) && (
          <label className="label-big-holder ">
            <PartialCheckbox
              className="mx-2"
              checked={
                countActive === todos.size
                  ? false
                  : countActive === 0
                  ? true
                  : undefined
              }
              onChange={({ target }) =>
                pushLoader(todoActions.toggleAll(!target.checked))
              }
            />
          </label>
        )}
        <InputGroup size="sm">
          <FormControl
            placeholder="Search ..."
            onChange={({ target }) => {
              filterActions.keyword(target.value.trimLeft());
            }}
          />
          {keyword ? (
            <InputGroup.Text
              as="button"
              className="btn-clear"
              onClick={handleClear}
            >
              <XLg />
            </InputGroup.Text>
          ) : (
            <InputGroup.Text
              as="button"
              className="btn-focus"
              onClick={handleFocus}
            >
              <Search />
            </InputGroup.Text>
          )}
        </InputGroup>
        <DropdownButton
          title={status}
          onSelect={(key) => filterActions.status(key as StatusFilter)}
          size="sm"
          className="mx-2"
          align="end"
        >
          {statusList.map((item) => {
            return (
              <Dropdown.Item key={item} eventKey={item}>
                {item}
              </Dropdown.Item>
            );
          })}
        </DropdownButton>
        <Button
          size="sm"
          onClick={handleDelete}
          variant="danger"
          disabled={todos.size === countActive}
        >
          <Trash />
        </Button>
        <LogoutButton />
      </Container>
    </Navbar>
  );
}
