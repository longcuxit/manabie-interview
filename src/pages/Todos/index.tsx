import { useEffect } from "react";
import { Container } from "react-bootstrap";

import { usePushLoader } from "components/Loader";
import { useTodoActions } from "store/Todos";

import TodoForm from "./Form";
import TodoList from "./List";
import TodoToolbar from "./Toolbar";

const ToDoPage = () => {
  const todoActions = useTodoActions();
  const pushLoader = usePushLoader();

  useEffect(() => {
    pushLoader(todoActions.initialize());
  }, [pushLoader, todoActions]);

  return (
    <div>
      <TodoToolbar />
      <TodoList />
      <div className="position-sticky bg-light py-2" style={{ bottom: 0 }}>
        <Container>
          <TodoForm onSubmit={todoActions.create} />
        </Container>
      </div>
    </div>
  );
};

export default ToDoPage;
