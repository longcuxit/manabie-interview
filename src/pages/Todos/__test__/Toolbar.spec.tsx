import { act, fireEvent, render } from "@testing-library/react";
import { TodoModel, TodoStatus } from "models/Todo";
import Item from "../Item";

import { useTodos } from "store/Todos";
import { useAsyncConfirm } from "components/AsyncModal";
import { usePushLoader } from "components/Loader";
import { useTodoFilter } from "../Store.filter";
import { PartialCheckbox } from "components/PartialCheckbox";

import TodoToolbar from "../Toolbar";
import { fakeTodo } from "utils/testing";

jest.mock("../Store.filter", () => ({ useTodoFilter: jest.fn() }));
jest.mock("store/Todos", () => ({ useTodos: jest.fn() }));
jest.mock("components/AsyncModal", () => ({ useAsyncConfirm: jest.fn() }));
jest.mock("components/PartialCheckbox", () => ({
  PartialCheckbox: jest.fn(() => <>PartialCheckbox</>),
}));
jest.mock("components/Loader", () => ({
  usePushLoader: jest.fn(),
  withLoader: (Com: any) => Com,
}));

const mockUseTodos = useTodos as jest.Mock;
const mockUseAsyncConfirm = useAsyncConfirm as jest.Mock;
const mockUsePushLoader = usePushLoader as jest.Mock;
const mockUseTodoFilter = useTodoFilter as jest.Mock;
const mockPartialCheckbox = PartialCheckbox as jest.Mock;

const mapTodos = new Map<string, TodoModel>();
Array.from({ length: 3 }, (_, i) => {
  const todo = fakeTodo({
    content: `Content-${i}`,
    status: i % 2 ? TodoStatus.COMPLETED : TodoStatus.ACTIVE,
  });
  mapTodos.set(todo.id, todo);
});

describe("pages/Todos/ToolBar: ", () => {
  const pushLoader = jest.fn();
  const updateStatus = jest.fn();
  const showConfirm = jest.fn();
  const clearAll = jest.fn();
  const filterStatus = jest.fn();
  const filterKeyword = jest.fn();
  const toggleAll = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();

    mockUseTodoFilter.mockReturnValue([
      { status: "ALL", keyword: "" },
      { status: filterStatus, keyword: filterKeyword },
    ]);
    mockUsePushLoader.mockReturnValue(pushLoader);
    mockUseTodos.mockReturnValue([
      { todos: mapTodos, countActive: 1 },
      { updateStatus, clearAll, toggleAll },
    ]);
    mockUseAsyncConfirm.mockReturnValue(showConfirm);
  });

  it("should match snapshot", () => {
    const { container } = render(<TodoToolbar />);

    expect(container).toMatchSnapshot();
  });

  describe("filter keyword: ", () => {
    it("should focus input when click focus button", async () => {
      const { container } = render(<TodoToolbar />);

      const focusBtn = container.querySelector(".btn-focus")!;
      const input = container.querySelector(
        ".form-control"
      ) as HTMLInputElement;
      const focus = (input.focus = jest.fn());

      fireEvent.click(focusBtn);

      expect(focus).toBeCalledTimes(1);
    });

    it("should filter when change value", async () => {
      const { container } = render(<TodoToolbar />);

      const input = container.querySelector(
        ".form-control"
      ) as HTMLInputElement;

      fireEvent.change(input, { target: { value: "Keyword" } });

      expect(filterKeyword).toBeCalledWith("Keyword");
    });

    it("should show clear button & reset value", async () => {
      mockUseTodoFilter.mockReturnValue([
        { status: "ALL", keyword: "Keyword" },
        { status: filterStatus, keyword: filterKeyword },
      ]);

      const { container } = render(<TodoToolbar />);

      const input = container.querySelector(
        ".form-control"
      ) as HTMLInputElement;
      const clearBtn = container.querySelector(".btn-clear")!;

      expect(clearBtn).toBeInTheDocument();

      input.value = "Keyword";

      fireEvent.click(clearBtn);

      expect(input.value).toBe("");
    });
  });

  describe("toggle all", () => {
    it("should show partial", () => {
      mockUseTodos.mockReturnValue([{ todos: mapTodos, countActive: 1 }]);

      render(<TodoToolbar />);

      expect(mockPartialCheckbox).toBeCalled();

      const props = mockPartialCheckbox.mock.calls[0][0];
      expect(props.checked).toBe(undefined);
    });

    it("should show checked & complete all todos", () => {
      mockUseTodos.mockReturnValue([
        { todos: mapTodos, countActive: 0 },
        { toggleAll },
      ]);

      render(<TodoToolbar />);

      expect(mockPartialCheckbox).toBeCalled();

      const props = mockPartialCheckbox.mock.calls[0][0];
      expect(props.checked).toBe(true);

      props.onChange({ target: { checked: false } });
      expect(toggleAll).toBeCalledWith(true);
    });

    it("should show unChecked & active all todos", () => {
      mockUseTodos.mockReturnValue([
        { todos: mapTodos, countActive: 3 },
        { toggleAll },
      ]);

      render(<TodoToolbar />);

      expect(mockPartialCheckbox).toBeCalled();

      const props = mockPartialCheckbox.mock.calls[0][0];
      expect(props.checked).toBe(false);

      props.onChange({ target: { checked: true } });
      expect(toggleAll).toBeCalledWith(false);
    });
  });

  describe("delete all ", () => {
    it("should show confirm % cancel delete", async () => {
      showConfirm.mockResolvedValue(undefined);

      const { container } = render(<TodoToolbar />);

      await act(async () => {
        fireEvent.click(container.querySelector(".btn-danger")!);
      });

      expect(showConfirm).toBeCalled();
      expect(pushLoader).not.toBeCalled();
      expect(clearAll).not.toBeCalled();
    });

    it("should show confirm % accept delete", async () => {
      showConfirm.mockResolvedValue(true);

      const { container } = render(<TodoToolbar />);

      await act(async () => {
        fireEvent.click(container.querySelector(".btn-danger")!);
      });

      expect(showConfirm).toBeCalled();
      expect(pushLoader).toBeCalled();
      expect(clearAll).toBeCalled();
    });

    it("should ignore show confirm", async () => {
      const { container, rerender } = render(
        <>
          <TodoToolbar />
          {false}
        </>
      );

      await act(async () => {
        fireEvent.click(container.querySelector(".btn-danger")!);
      });

      expect(showConfirm).toBeCalled();

      rerender(
        <>
          <TodoToolbar />
          {showConfirm.mock.calls[0][0].footer}
        </>
      );

      const askMe = container.querySelector(".flex-fill .form-check-input")!;

      expect(askMe).toBeInTheDocument();
      expect(localStorage.getItem("DeleteAllItems")).toBe("true");
      fireEvent.click(askMe);
      expect(localStorage.getItem("DeleteAllItems")).toBe("false");
    });

    it("should don't show confirm % accept delete", async () => {
      showConfirm.mockResolvedValue(true);

      const { container } = render(<TodoToolbar />);

      await act(async () => {
        fireEvent.click(container.querySelector(".btn-danger")!);
      });

      expect(showConfirm).not.toBeCalled();
      expect(pushLoader).toBeCalled();
      expect(clearAll).toBeCalled();
    });
  });

  it("filter status", () => {
    const { container } = render(<TodoToolbar />);

    fireEvent.click(container.querySelector(".dropdown-toggle")!);

    expect(container.querySelector(".dropdown-menu")).toBeInTheDocument();

    fireEvent.click(container.querySelectorAll(".dropdown-item")[1]);

    expect(filterStatus).toBeCalledWith("ACTIVE");
  });
});
