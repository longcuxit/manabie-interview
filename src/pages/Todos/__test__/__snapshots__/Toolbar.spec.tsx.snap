// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/Todos/ToolBar:  should match snapshot 1`] = `
<div>
  <nav
    class="sticky-top navbar navbar-expand navbar-light bg-primary"
  >
    <div
      class="container"
    >
      <div
        class="input-group input-group-sm"
      >
        <input
          class="form-control"
          placeholder="Search ..."
        />
        <button
          class="btn-focus input-group-text"
        >
          <svg
            fill="currentColor"
            height="1em"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11.742 10.344a6.5 6.5 0 1 0-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 0 0 1.415-1.414l-3.85-3.85a1.007 1.007 0 0 0-.115-.1zM12 6.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0z"
            />
          </svg>
        </button>
      </div>
      <div
        class="mx-2 dropdown"
      >
        <button
          aria-expanded="false"
          class="dropdown-toggle btn btn-primary btn-sm"
          type="button"
        >
          ALL
        </button>
      </div>
      <button
        class="btn btn-danger btn-sm"
        type="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5.5 5.5A.5.5 0 0 1 6 6v6a.5.5 0 0 1-1 0V6a.5.5 0 0 1 .5-.5zm2.5 0a.5.5 0 0 1 .5.5v6a.5.5 0 0 1-1 0V6a.5.5 0 0 1 .5-.5zm3 .5a.5.5 0 0 0-1 0v6a.5.5 0 0 0 1 0V6z"
          />
          <path
            d="M14.5 3a1 1 0 0 1-1 1H13v9a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V4h-.5a1 1 0 0 1-1-1V2a1 1 0 0 1 1-1H6a1 1 0 0 1 1-1h2a1 1 0 0 1 1 1h3.5a1 1 0 0 1 1 1v1zM4.118 4 4 4.059V13a1 1 0 0 0 1 1h6a1 1 0 0 0 1-1V4.059L11.882 4H4.118zM2.5 3V2h11v1h-11z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      <button
        class="ms-2 btn btn-primary btn-sm"
        type="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10 12.5a.5.5 0 0 1-.5.5h-8a.5.5 0 0 1-.5-.5v-9a.5.5 0 0 1 .5-.5h8a.5.5 0 0 1 .5.5v2a.5.5 0 0 0 1 0v-2A1.5 1.5 0 0 0 9.5 2h-8A1.5 1.5 0 0 0 0 3.5v9A1.5 1.5 0 0 0 1.5 14h8a1.5 1.5 0 0 0 1.5-1.5v-2a.5.5 0 0 0-1 0v2z"
            fill-rule="evenodd"
          />
          <path
            d="M15.854 8.354a.5.5 0 0 0 0-.708l-3-3a.5.5 0 0 0-.708.708L14.293 7.5H5.5a.5.5 0 0 0 0 1h8.793l-2.147 2.146a.5.5 0 0 0 .708.708l3-3z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
  </nav>
</div>
`;
